lib/engine/pa_messages.ex

```diff
@@ -0,0 +1,191 @@
+  require Logger
+
+  alias PaMessages.PaMessage
+
+  @type state :: %{
+          pa_message_timers_table: :ets.tab()
+        }
+
+  @pa_message_timers_table :pa_message_timers
+  @minute_in_ms 1000 * 60
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  @impl true
+  def init(_) do
+    # Add some delay to wait for sign processes to come up before sending PA messages
+    Process.send_after(self(), :update, 10000)
+    state = %{pa_message_timers_table: @pa_message_timers_table}
+    create_table(state)
+    {:ok, state}
+  end
+
+  def create_table(state) do
+    :ets.new(state.pa_message_timers_table, [:named_table, read_concurrency: true])
+  end
+
+  @impl true
+  def handle_info(:update, state) do
+    schedule_update(self())
+
+    case get_active_pa_messages() do
+      {:ok, pa_messages} ->
+        schedule_pa_messages(pa_messages, state.pa_message_timers_table)
+        |> handle_inactive_pa_messages(state.pa_message_timers_table)
+
+      {:error, %HTTPoison.Response{status_code: status_code, body: body}} ->
+        Logger.error("pa_messages_response_error: status_code=#{status_code} body=#{body}")
+
+      {:error, %HTTPoison.Error{reason: reason}} ->
+        Logger.error("pa_messages_response_error: reason=#{reason}")
+
+      {:error, error} ->
+        Logger.error("pa_messages_response_error: error=#{inspect(error)}")
+    end
+
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info({:send_pa_message, pa_message}, state) do
+    Enum.each(pa_message.sign_ids, fn sign_id ->
+      send(
+        String.to_existing_atom("Signs/#{sign_id}"),
+        {:play_pa_message, pa_message}
+      )
+    end)
+
+    {:noreply, state}
+  end
+
+  defp schedule_pa_messages(pa_messages, table) do
+    for %{
+          "id" => pa_id,
+          "visual_text" => visual_text,
+          "audio_text" => audio_text,
+          "interval_in_minutes" => interval_in_minutes,
+          "priority" => priority,
+          "sign_ids" => sign_ids
+        } <- pa_messages do
+      active_pa_message = %PaMessage{
+        id: pa_id,
+        visual_text: visual_text,
+        audio_text: audio_text,
+        priority: priority,
+        sign_ids: sign_ids,
+        interval_in_ms: interval_in_minutes * @minute_in_ms
+      }
+
+      case get_pa_message_timer(pa_id, table) do
+        {timer_ref, existing_pa_message}
+        when existing_pa_message.interval_in_ms != active_pa_message.interval_in_ms ->
+          case Process.read_timer(timer_ref) do
+            false ->
+              schedule_pa_message(active_pa_message, active_pa_message.interval_in_ms, table)
+
+            remaining_ms ->
+              ms_elapsed = existing_pa_message.interval_in_ms - remaining_ms
+              temp_interval = (active_pa_message.interval_in_ms - ms_elapsed) |> max(0)
+              cancel_pa_timer(timer_ref, pa_id)
+              schedule_pa_message(active_pa_message, temp_interval, table)
```

Comment:
Hmm yeah I think that approach works nicely and will make the "edited interval" piece of the logic a good deal similar. And it removes the need to cancel timers and such as you hinted at. I think we also would get the "handling" of inactive messages for free as well since if they're not included in the response, we won't be checking for the last playtime anyway.